src/import/song.rs

```diff
@@ -0,0 +1,87 @@
+use std::path::Path;
+use taglib2_sys::SongProperties;
+
+use chrono::prelude::*;
+
+use database::song;
+use database::song_artist;
+use diesel;
+use diesel::Connection;
+use diesel::prelude::*;
+use diesel::result;
+
+use models::*;
+
+use import::album::add_or_get_album;
+use import::artist::add_or_get_artist;
+
+use super::errors;
+
+/// Takes information about a song read from tags and adds it to the database.
+pub fn add_song(
+    path: &Path,
+    artwork_directory: &Path,
+    props: SongProperties,
+    conn: &SqliteConnection,
+) -> super::errors::Result<()> {
+    let artist: Option<Artist> = props
+        .artist
+        .as_ref()
+        .map_or(Ok(None), |name| add_or_get_artist(name, conn).map(Some))?;
+
+    let album_artist: Option<Artist> = props
+        .album_artist
+        .as_ref()
+        .map_or(Ok(None), |name| add_or_get_artist(name, conn).map(Some))?;
+
+    let (artist, album_artist) = (
+        artist
+            .clone()
+            .or(album_artist.clone())
+            .ok_or(errors::ErrorKind::NoArtistError)?,
+        album_artist
+            .or(artist)
+            .ok_or(errors::ErrorKind::NoArtistError)?,
+    );
+
+    let album_name = props.album.as_ref().ok_or(errors::ErrorKind::NoAlbumError)?;
```

Comment:
This can probably happen as part of `add_or_get_album` so that you don't pass in more than you need. You only need to pass in the song properties, not the album name and year as well because you can already get those off of the song properties.